modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,438 @@
+         * q^p = e^{pln(q)}
+         */
+        template <typename T>
+        friend Quat<T> power(const Quat<T> &q, const Quat<T> &p);
+
+        Quat<_Tp> power(const Quat<_Tp> &p) const;
+        /**
+         * @ brief return the crossProduct between q and q1
+         */
+        template <typename T>
+        friend Quat<T> crossProduct(const Quat<T> &p, const Quat<T> &q);
+
+        Quat<_Tp> crossProduct(const Quat<_Tp> &q) const;
+
+        /**
+         * @brief return the norm of quaternion
+         * ||q|| = \sqrt{w^2 + x^2 + y^2 + z^2}
+         */
+        _Tp norm() const;
+
+        /**
+         * @brief return a normalzed q
+         * q_n = \frac{q}{||q||}
+         * where q_n.i satisfy q_n.x^2 + q_n.y^2 + q_n.z^2 + q_n.w^2 = 1
+         */
+        Quat<_Tp> normalize() const;
+
+        /**
+         * @brief return an inverse q q^-1
+         * which satisfies q * q^-1 = 1
+         */
+        template <typename T>
+        friend Quat<T> inv(const Quat<T> &q1);
+        Quat<_Tp> inv() const;
+
+        /**
+         * @brief sinh(p) = sin(w)cos(||v||) + cosh(w)\frac{v}{||v||}sin||v||
+         */
+        /*
+        template <typename T>
+        friend Quat<T> sinh(const Quat<T> &q1);
+        */
+        
+	Quat<_Tp> sinh() const;
+
+/*
+        template <typename T>
+        friend Quat<T> cosh(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> cosh() const;
+
+/*
+        template <typename T>
+        friend Quat<T> tanh(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> tanh() const;
+
+/*
+        template <typename T>
+        friend Quat<T> sin(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> sin() const;
+
+/*
+        template <typename T>
+        friend Quat<T> cos(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> cos() const;
+
+/*
+        template <typename T>
+        friend Quat<T> tan(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> tan() const;
+
+/*
+        template <typename T>
+        friend Quat<T> asin(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> asin() const;
+
+/*
+        template <typename T>
+        friend Quat<T> acos(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> acos() const;
+
+/*
+        template <typename T>
+        friend Quat<T> atan(const Quat<T> &q1);
+*/
+
+        Quat<_Tp> atan() const;
+/*
+        template <typename T>
+        friend Quat<T> asinh(const Quat<T> &q1);
+*/
+        Quat<_Tp> asinh() const;
+/*
+        template <typename T>
+        friend Quat<T> acosh(const Quat<T> &q1);
+*/
+        Quat<_Tp> acosh() const;
+
+        //template <typename T>
+        //friend Quat<T> atanh(const Quat<T> &q1);
+
+        Quat<_Tp> atanh() const;
+
+        /**
+         * @brirf to dermined whether a quaternion is normalized or not
+         */
+        //bool isNormal(_Tp esp=defaultVa) const;
+        bool isNormal() const;
+
+        /**
+         * @brief to throw an un-normalized error if its not an unit-quaternion
+         */
+        void assertNormal() const;
```

Comment:
As I said in earlier discussion, norm checking involves comparison of a (norm()-1) with an epsilon value which can be different in a source of quaternions. There may be an unpleasant situation when a user function produces unit quaternions and our implementation considers them not unit enough.
I propose to get rid of such functions or to add an epsilon argument.

Reply:
Ok, thanks. It has been dropped in my latest code.

Reply:
After discussion with Longbu, we think the assertNormal could remain to use for users with a param eps. They could check if a quaternion is a unit quaternion within the eps range that they can bear. 
What do you think?

Reply:
Agree, this is a good solution